google/cloud/pubsub/internal/tracing_batch_callback.cc

```diff
@@ -137,16 +139,51 @@ class TracingBatchCallback : public BatchCallback {
   }
 
  private:
+  struct MessageSpans {
+    opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> subscribe_span;
+    opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span>
+        concurrency_control_span;
+  };
+
+  void StartConcurrencyControl(std::string const& ack_id) override {
+    namespace sc = opentelemetry::trace::SemanticConventions;
+    std::lock_guard<std::mutex> lk(mu_);
+    auto spans = spans_by_ack_id_.find(ack_id);
+    if (spans != spans_by_ack_id_.end()) {
+      auto subscribe_span = spans->second.subscribe_span;
+      if (subscribe_span) {
+        opentelemetry::trace::StartSpanOptions options;
+        options.parent = subscribe_span->GetContext();
+        auto span =
+            internal::MakeSpan("subscriber concurrency control",
+                               {{sc::kMessagingSystem, "gcp_pubsub"}}, options);
+        auto scope = internal::OTelScope(span);
+        spans->second.concurrency_control_span = span;
+      }
+    }
+  }
+
+  void EndConcurrencyControl(std::string const& ack_id) override {
+    std::lock_guard<std::mutex> lk(mu_);
+    auto spans = spans_by_ack_id_.find(ack_id);
+    if (spans != spans_by_ack_id_.end()) {
+      auto concurrency_control_span = spans->second.concurrency_control_span;
+      if (concurrency_control_span) {
+        concurrency_control_span->End();
+      }
+    }
+  }
+
   void AddEvent(std::string const& ack_id, std::string const& event,
                 bool end_event = false) {
     std::lock_guard<std::mutex> lk(mu_);
     // Use the ack_id to find the subscribe span and add an event to it.
-    auto subscribe_span = subscribe_span_by_ack_id_.find(ack_id);
-    if (subscribe_span != subscribe_span_by_ack_id_.end()) {
-      subscribe_span->second->AddEvent(event);
+    auto spans = spans_by_ack_id_.find(ack_id);
+    if (spans != spans_by_ack_id_.end()) {
+      spans->second.subscribe_span->AddEvent(event);
       if (end_event) {
-        subscribe_span->second->End();
-        subscribe_span_by_ack_id_.erase(subscribe_span);
+        spans->second.subscribe_span->End();
```

Comment:
Q. What is the state of `spans->second.concurrency_control_span` here?

Reply:
It should have been ended at this point and now null (since we moved it above)

Reply:
Q.  "Should" or "will"?  That is, could you assert it?  (I'm not suggesting that you do ... just asking if you could.)

Reply:
It will have ended if an ack/nack comes after the user callback completes. 

If a message expires, I'm not sure. I think it will not have ended so we should handle that here in case.

Reply:
Actually, nevermind. When it expires, it does not end the span until the nack is called.